core/wasm/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+        /// This function doesn't take ownership of the provided allocation
+        /// but it does transfer ownership of the result to the caller
+        pub(super) fn iroha_wasm_execute_query(ptr: usize, len: usize) -> usize;
+
+        /// Executes encoded instruction by providing offset and length
+        /// into WebAssembly's linear memory where instruction is stored
+        ///
+        /// # Warning
+        ///
+        /// This function doesn't take ownership of the provided allocation
+        /// but it does transfer ownership of the result to the caller
+        pub(super) fn iroha_wasm_execute_instruction(ptr: usize, len: usize);
+    }
+}
+
+/// Decode the object from given pointer where first element is the size of the object
+/// following it. This can be considered a custom encoding format.
+///
+/// # Warning
+///
+/// This method takes ownership of the given pointer
+///
+/// # Safety
+///
+/// It's safe to call this function as long as it's safe to construct, from the given
+/// pointer, byte array of prefix length and `Box<[u8]>` containing the encoded object
+unsafe fn decode_with_length_prefix_from_raw<T: Decode>(ptr: usize) -> T {
+    let len_size_bytes = core::mem::size_of::<usize>();
+
+    #[allow(clippy::expect_used)]
+    let len = usize::from_be_bytes(
+        core::slice::from_raw_parts(ptr as *mut _, len_size_bytes)
+            .try_into()
+            .expect("Prefix length size(bytes) incorrect. This is a bug."),
+    );
+
+    _decode_from_raw_in_range(ptr, len, len_size_bytes..)
+}
+
+/// Decode the object from given pointer and length
+///
+/// # Warning
+///
+/// This method takes ownership of the given pointer
+///
+/// # Safety
+///
+/// It's safe to call this function as long as it's safe to construct, from the given
+/// pointer, `Box<[u8]>` containing the encoded object
+pub unsafe fn _decode_from_raw<T: Decode>(ptr: usize, len: usize) -> T {
+    _decode_from_raw_in_range(ptr, len, 0..)
+}
+
+/// Decode the object from given pointer and length in the given range
+///
+/// # Warning
+///
+/// This method takes ownership of the given pointer
+///
+/// # Safety
+///
+/// It's safe to call this function as long as it's safe to construct, from the given
+/// pointer, `Box<[u8]>` containing the encoded object
+// `usize` is always pointer sized
+#[allow(clippy::cast_possible_truncation)]
+unsafe fn _decode_from_raw_in_range<T: Decode>(
+    ptr: usize,
+    len: usize,
+    range: RangeFrom<usize>,
+) -> T {
+    let bytes = Box::from_raw(core::slice::from_raw_parts_mut(ptr as *mut _, len));
+
+    #[allow(clippy::expect_used, clippy::expect_fun_call)]
+    T::decode(&mut &bytes[range]).expect(
+        format!(
+            "Decoding of {} failed. This is a bug",
+            core::any::type_name::<T>()
+        )
+        .as_str(),
+    )
+}
+
+#[no_mangle]
+/// Multiplexed all functions
+unsafe extern "C" fn _iroha_wasm_fn(fn_id: u32, args_ptr: usize, args_len: usize) -> usize {
+    42
+}
```

Comment:
So you are saying, that using functions for all iroha functions will increase wasm binary size?

Maybe we can you feature flags for different interfaces? One with functions and increased binary size and one with this tiny implementation?

I just believe that for user it's better to try simple interface at first and do some prototyping before using this hard optimizations